File: LOL02/LOL02/LOL02/LOL/Physics.cs
/**
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <http://unlicense.org>
 */

using System;

using Microsoft.Xna.Framework;

using FarseerPhysics.Collision;
using FarseerPhysics.Common;
using FarseerPhysics.Dynamics;
using FarseerPhysics.Dynamics.Contacts;
using FarseerPhysics.Dynamics.Joints;
using FarseerPhysics.Factories;

namespace LOL
{
    /**
     * Game designers can configure the physics of a level (i.e., the default
     * forces, if any), via this class. Internally, the class constructs a box2d
     * physics world and instantiates the callbacks needed to ensure that LOL works
     * correctly.
     */
    public class Physics
    {
        /**
         * This ratio means that every 10 pixels on the screen will correspond to a
         * meter. Note that 'pixels' are defined in terms of what a programmer's
         * Config says, not the actual screen size, because the programmer's Config
         * gets scaled to screen dimensions.
         */
        internal static float PIXEL_METER_RATIO = 10;

        /**
         * When a hero collides with a "sticky" obstacle, this is the code we run to
         * figure out what to do
         * 
         * @param sticky The sticky entity... it should always be an obstacle for
         *            now
         * @param other The other entity... it should always be a hero for now
         * @param contact A description of the contact event
         */
        internal static void HandleSticky(PhysicsSprite sticky, PhysicsSprite other, Contact contact)
        {
            if (other.mDJoint != null) { return; }
            if (other.mStickyDelay != null && DateTime.Now < other.mStickyDelay)
            {
                return;
            }
            if ((sticky.mIsSticky[0] && other.YPosition >= sticky.YPosition + sticky.mSize.Y)
                || (sticky.mIsSticky[1] && other.XPosition + other.mSize.X <= sticky.XPosition)
                || (sticky.mIsSticky[3] && other.XPosition >= sticky.XPosition + sticky.mSize.X)
                || (sticky.mIsSticky[2] && other.YPosition + other.mSize.Y <= sticky.YPosition))
            {
                Vector2 normal;
                FixedArray2<Vector2> points;
                contact.GetWorldManifold(out normal, out points);
                Vector2 v = points[0];
                Level.sCurrent.mOneTimeEvents.Add(() =>
                {
                    other.mBody.LinearVelocity = new Vector2(0, 0);
                    DistanceJoint d = JointFactory.CreateDistanceJoint(Level.sCurrent.mWorld, sticky.mBody, other.mBody, v, v, false);
                    d.CollideConnected = true;
                    other.mDJoint = d;

                    WeldJoint w = JointFactory.CreateWeldJoint(Level.sCurrent.mWorld, sticky.mBody, other.mBody, v, v, false);
                    w.CollideConnected = true;
                    other.mWJoint = w;
                });

            }
        }

        /**
         * Configure physics for the current level
         * 
         * @param defaultXGravity The default force moving entities to the left
         *            (negative) or right (positive)... Usually zero
         * @param defaultYGravity The default force pushing the hero down (negative)
         *            or up (positive)... Usually zero or -10
         */
        public static void Configure(float defaultXGravity, float defaultYGravity)
        {
            Level.sCurrent.mWorld = new World(new Vector2(defaultXGravity, defaultYGravity));
            BeginContactDelegate beginContact = delegate(Contact contact)
            {
                Object a = contact.FixtureA.Body.UserData;
                Object b = contact.FixtureB.Body.UserData;

                if (!(a is PhysicsSprite) || !(b is PhysicsSprite)) { return false; }
                PhysicsSprite c0, c1;
                if (a is Hero)
                {
                    c0 = (PhysicsSprite)a;
                    c1 = (PhysicsSprite)b;
                }
                else if (b is Hero)
                {
                    c0 = (PhysicsSprite)b;
                    c1 = (PhysicsSprite)a;
                }
                else if (a is Enemy)
                {
                    c0 = (PhysicsSprite)a;
                    c1 = (PhysicsSprite)b;
                }
                else if (b is Enemy)
                {
                    c0 = (PhysicsSprite)b;
                    c1 = (PhysicsSprite)a;
                }
                else if (a is Projectile)
                {
                    c0 = (PhysicsSprite)a;
                    c1 = (PhysicsSprite)b;
                }
                else if (b is Projectile)
                {
                    c0 = (PhysicsSprite)b;
                    c1 = (PhysicsSprite)a;
                }
                else
                {
                    return true;
                }
                Level.sCurrent.mOneTimeEvents.Add(delegate()
                {
                    c0.OnCollide(c1, contact);
                });
                return true;
            };
            PreSolveDelegate preSolve = delegate(Contact contact, ref Manifold oldManifold)
            {
                Object a = contact.FixtureA.Body.UserData;
                Object b = contact.FixtureB.Body.UserData;

                if (!(a is PhysicsSprite) || !(b is PhysicsSprite)) { return; }

                PhysicsSprite gfoA = (PhysicsSprite)a;
                PhysicsSprite gfoB = (PhysicsSprite)b;

                if (gfoA.mIsSticky[0] || gfoA.mIsSticky[1] || gfoA.mIsSticky[2] || gfoA.mIsSticky[3])
                {
                    HandleSticky(gfoA, gfoB, contact);
                    return;
                }
                else if (gfoB.mIsSticky[0] || gfoB.mIsSticky[1] || gfoB.mIsSticky[2] || gfoB.mIsSticky[3])
                {
                    HandleSticky(gfoB, gfoA, contact);
                    return;
                }

                if (gfoA.mPassThroughId != 0 && gfoA.mPassThroughId == gfoB.mPassThroughId)
                {
                    contact.Enabled = false;
                    return;
                }

                PhysicsSprite onesided = null;
                PhysicsSprite other = null;

                if (gfoA.mIsOneSided > -1)
                {
                    onesided = gfoA;
                    other = gfoB;
                }
                else if (gfoB.mIsOneSided > -1)
                {
                    onesided = gfoB;
                    other = gfoA;
                }
                else { return; }

                Vector2 normal;
                FixedArray2<Vector2> points;
                contact.GetWorldManifold(out normal, out points);
                for (int i = 0; i < 2; ++i)
                {
                    Vector2 v2 = points[i];
                    if (onesided.mIsOneSided == 0 && v2.Y < 0)
                    {
                        contact.Enabled = false;
                    }
                    else if (onesided.mIsOneSided == 2 && v2.Y > 0)
                    {
                        contact.Enabled = false;
                    }
                    else if (onesided.mIsOneSided == 1 && v2.X > 0)
                    {
                        contact.Enabled = false;
                    }
                    else if (onesided.mIsOneSided == 3 && v2.X < 0)
                    {
                        contact.Enabled = false;
                    }
                }
            };

            Level.sCurrent.mWorld.ContactManager.BeginContact += beginContact;
            Level.sCurrent.mWorld.ContactManager.PreSolve += preSolve;
        }
    }
}

File: core/src/edu/lehigh/cse/lol/Physics.java
/**
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <http://unlicense.org>
 */

package edu.lehigh.cse.lol;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.Contact;
import com.badlogic.gdx.physics.box2d.ContactImpulse;
import com.badlogic.gdx.physics.box2d.ContactListener;
import com.badlogic.gdx.physics.box2d.Manifold;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.physics.box2d.WorldManifold;
import com.badlogic.gdx.physics.box2d.joints.DistanceJoint;
import com.badlogic.gdx.physics.box2d.joints.DistanceJointDef;
import com.badlogic.gdx.physics.box2d.joints.WeldJoint;
import com.badlogic.gdx.physics.box2d.joints.WeldJointDef;

import edu.lehigh.cse.lol.internals.LolAction;

/**
 * Game designers can configure the physics of a level (i.e., the default
 * forces, if any), via this class. Internally, the class constructs a box2d
 * physics world and instantiates the callbacks needed to ensure that LOL works
 * correctly.
 */
public class Physics {
    /**
     * This ratio means that every 20 pixels on the screen will correspond to a
     * meter. Note that 'pixels' are defined in terms of what a programmer's
     * configuration() says, not the actual screen size, because the
     * configuration gets scaled to screen dimensions. The default is 960x640.
     */
    static final float PIXEL_METER_RATIO = 20;

    /**
     * When a hero collides with a "sticky" obstacle, this is the code we run to
     * figure out what to do
     *
     * @param sticky  The sticky actor... it should always be an obstacle for now
     * @param other   The other actor... it should always be a hero for now
     * @param contact A description of the contact event
     */
    static void handleSticky(final Actor sticky, final Actor other, Contact contact) {
        // don't create a joint if we've already got one
        if (other.mDJoint != null)
            return;
        // don't create a joint if we're supposed to wait
        if (System.currentTimeMillis() < other.mStickyDelay)
            return;
        // handle sticky obstacles... only do something if we're hitting the
        // obstacle from the correct direction
        if ((sticky.mIsSticky[0] && other.getYPosition() >= sticky.getYPosition() + sticky.mSize.y)
                || (sticky.mIsSticky[1] && other.getXPosition() + other.mSize.x <= sticky.getXPosition())
                || (sticky.mIsSticky[3] && other.getXPosition() >= sticky.getXPosition() + sticky.mSize.x)
                || (sticky.mIsSticky[2] && other.getYPosition() + other.mSize.y <= sticky.getYPosition())) {
            // create distance and weld joints... somehow, the combination is
            // needed to get this to work. Note that this function runs during
            // the box2d step, so we need to make the joint in a callback that
            // runs later
            final Vector2 v = contact.getWorldManifold().getPoints()[0];
            Lol.sGame.mCurrentLevel.mOneTimeEvents.add(new LolAction() {
                @Override
                public void go() {
                    other.mBody.setLinearVelocity(0, 0);
                    DistanceJointDef d = new DistanceJointDef();
                    d.initialize(sticky.mBody, other.mBody, v, v);
                    d.collideConnected = true;
                    other.mDJoint = (DistanceJoint) Lol.sGame.mCurrentLevel.mWorld.createJoint(d);
                    WeldJointDef w = new WeldJointDef();
                    w.initialize(sticky.mBody, other.mBody, v);
                    w.collideConnected = true;
                    other.mWJoint = (WeldJoint) Lol.sGame.mCurrentLevel.mWorld.createJoint(w);
                }
            });
        }
    }

    /**
     * Configure physics for the current level
     *
     * @param defaultXGravity The default force moving actors to the left (negative) or
     *                        right (positive)... Usually zero
     * @param defaultYGravity The default force pushing actors down (negative) or up
     *                        (positive)... Usually zero or -10
     */
    public static void configure(float defaultXGravity, float defaultYGravity) {
        // create a world with gravity
        Lol.sGame.mCurrentLevel.mWorld = new World(new Vector2(defaultXGravity, defaultYGravity), true);

        // set up the collision handlers
        Lol.sGame.mCurrentLevel.mWorld.setContactListener(new ContactListener() {
            /**
             * When two bodies start to collide, we can use this to forward to
             * our onCollide methods
             */
            @Override
            public void beginContact(final Contact contact) {
                // Get the bodies, make sure both are actors
                Object a = contact.getFixtureA().getBody().getUserData();
                Object b = contact.getFixtureB().getBody().getUserData();
                if (!(a instanceof Actor) || !(b instanceof Actor))
                    return;

                // the order is Hero, Enemy, Goodie, Projectile, Obstacle, Destination
                //
                // Of those, Hero, Enemy, and Projectile are the only ones with
                // a non-empty onCollide
                final Actor c0;
                final Actor c1;
                if (a instanceof Hero) {
                    c0 = (Actor) a;
                    c1 = (Actor) b;
                } else if (b instanceof Hero) {
                    c0 = (Actor) b;
                    c1 = (Actor) a;
                } else if (a instanceof Enemy) {
                    c0 = (Actor) a;
                    c1 = (Actor) b;
                } else if (b instanceof Enemy) {
                    c0 = (Actor) b;
                    c1 = (Actor) a;
                } else if (a instanceof Projectile) {
                    c0 = (Actor) a;
                    c1 = (Actor) b;
                } else if (b instanceof Projectile) {
                    c0 = (Actor) b;
                    c1 = (Actor) a;
                } else {
                    return;
                }

                // Schedule an event to run as soon as the physics world
                // finishes its step.
                //
                // NB: this is called from render, while world is updating...
                // you can't modify the world or its actors until the update
                // finishes, so we have to schedule collision-based updates to
                // run after the world update.
                Lol.sGame.mCurrentLevel.mOneTimeEvents.add(new LolAction() {
                    @Override
                    public void go() {
                        c0.onCollide(c1, contact);
                    }
                });
            }

            /**
             * We ignore endcontact
             */
            @Override
            public void endContact(Contact contact) {
            }

            /**
             * Presolve is a hook for disabling certain collisions. We use it
             * for collision immunity, sticky obstacles, and one-way walls
             */
            @Override
            public void preSolve(Contact contact, Manifold oldManifold) {
                // get the bodies, make sure both are actors
                Object a = contact.getFixtureA().getBody().getUserData();
                Object b = contact.getFixtureB().getBody().getUserData();
                if (!(a instanceof Actor) || !(b instanceof Actor))
                    return;
                Actor gfoA = (Actor) a;
                Actor gfoB = (Actor) b;

                // handle sticky obstacles... only do something if at least one
                // actor is a sticky actor
                if (gfoA.mIsSticky[0] || gfoA.mIsSticky[1] || gfoA.mIsSticky[2] || gfoA.mIsSticky[3]) {
                    handleSticky(gfoA, gfoB, contact);
                    return;
                } else if (gfoB.mIsSticky[0] || gfoB.mIsSticky[1] || gfoB.mIsSticky[2] || gfoB.mIsSticky[3]) {
                    handleSticky(gfoB, gfoA, contact);
                    return;
                }

                // if the actors have the same passthrough ID, and it's
                // not zero, then disable the contact
                if (gfoA.mPassThroughId != 0 && gfoA.mPassThroughId == gfoB.mPassThroughId) {
                    contact.setEnabled(false);
                    return;
                }

                // is either one-sided? If not, we're done
                Actor onesided = null;
                Actor other;
                if (gfoA.mIsOneSided > -1) {
                    onesided = gfoA;
                    other = gfoB;
                } else if (gfoB.mIsOneSided > -1) {
                    onesided = gfoB;
                    other = gfoA;
                } else {
                    return;
                }

                // if we're here, see if we should be disabling a one-sided
                // obstacle collision
                WorldManifold worldManiFold = contact.getWorldManifold();
                int numPoints = worldManiFold.getNumberOfContactPoints();
                for (int i = 0; i < numPoints; i++) {
                    Vector2 vector2 = other.mBody.getLinearVelocityFromWorldPoint(worldManiFold.getPoints()[i]);
                    // disable based on the value of isOneSided and the vector
                    // between the actors
                    if (onesided.mIsOneSided == 0 && vector2.y < 0)
                        contact.setEnabled(false);
                    else if (onesided.mIsOneSided == 2 && vector2.y > 0)
                        contact.setEnabled(false);
                    else if (onesided.mIsOneSided == 1 && vector2.x > 0)
                        contact.setEnabled(false);
                    else if (onesided.mIsOneSided == 3 && vector2.x < 0)
                        contact.setEnabled(false);
                }
            }

            /**
             * We ignore postsolve
             */
            @Override
            public void postSolve(Contact contact, ContactImpulse impulse) {
            }
        });
    }
}
